File: backend/Minimarket.Api/Controllers/ReportsController.cs
using Microsoft.AspNetCore.Mvc;
using Minimarket.Api.DTOs;
using Minimarket.Api.Services;

namespace Minimarket.Api.Controllers;

[ApiController]
[Route("api/[controller]")]
public class ReportsController(IReportService reportService) : ControllerBase
{
    [HttpGet("sales-summary")]
    public async Task<ActionResult<IReadOnlyCollection<SalesSummaryDto>>> GetSalesSummary(
        [FromQuery] DateTime? startDate,
        [FromQuery] DateTime? endDate)
    {
        var start = startDate?.Date ?? DateTime.Today.AddDays(-6);
        var end = endDate?.Date.AddDays(1).AddTicks(-1) ?? DateTime.Today.AddDays(1).AddTicks(-1);

        return Ok(await reportService.GetSalesSummaryAsync(start, end));
    }

    [HttpGet("top-products")]
    public async Task<ActionResult<IReadOnlyCollection<TopSellingProductDto>>> GetTopProducts(
        [FromQuery] DateTime? startDate,
        [FromQuery] DateTime? endDate,
        [FromQuery] int? limit)
    {
        var start = startDate?.Date ?? DateTime.Today.AddDays(-6);
        var end = endDate?.Date.AddDays(1).AddTicks(-1) ?? DateTime.Today.AddDays(1).AddTicks(-1);
        var topLimit = limit is > 0 and <= 20 ? limit.Value : 5;

        return Ok(await reportService.GetTopSellingProductsAsync(start, end, topLimit));
    }

    [HttpGet("dashboard")]
    public async Task<ActionResult<DashboardDto>> GetDashboard()
        => Ok(await reportService.GetDashboardAsync());
}

File: backend-java/minimarket-api-java/src/main/java/com/minimarket/api/controller/ReportsController.java
package com.minimarket.api.controller;

import com.minimarket.api.dto.DashboardDto;
import com.minimarket.api.dto.SalesSummaryDto;
import com.minimarket.api.dto.TopSellingProductDto;
import com.minimarket.api.service.ReportService;
import org.springframework.format.annotation.DateTimeFormat;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RestController;

import java.time.LocalDate;
import java.util.List;

@RestController
@RequestMapping("/api/reports")
public class ReportsController {

    private final ReportService reportService;

    public ReportsController(ReportService reportService) {
        this.reportService = reportService;
    }

    @GetMapping("/sales-summary")
    public List<SalesSummaryDto> getSalesSummary(
        @RequestParam(required = false) @DateTimeFormat(iso = DateTimeFormat.ISO.DATE) LocalDate startDate,
        @RequestParam(required = false) @DateTimeFormat(iso = DateTimeFormat.ISO.DATE) LocalDate endDate
    ) {
        var start = startDate != null ? startDate.atStartOfDay() : LocalDate.now().minusDays(6).atStartOfDay();
        var end = endDate != null
            ? endDate.plusDays(1).atStartOfDay().minusNanos(1)
            : LocalDate.now().plusDays(1).atStartOfDay().minusNanos(1);

        return reportService.getSalesSummary(start, end);
    }

    @GetMapping("/top-products")
    public List<TopSellingProductDto> getTopProducts(
        @RequestParam(required = false) @DateTimeFormat(iso = DateTimeFormat.ISO.DATE) LocalDate startDate,
        @RequestParam(required = false) @DateTimeFormat(iso = DateTimeFormat.ISO.DATE) LocalDate endDate,
        @RequestParam(required = false) Integer limit
    ) {
        var start = startDate != null ? startDate.atStartOfDay() : LocalDate.now().minusDays(6).atStartOfDay();
        var end = endDate != null
            ? endDate.plusDays(1).atStartOfDay().minusNanos(1)
            : LocalDate.now().plusDays(1).atStartOfDay().minusNanos(1);
        var topLimit = limit != null && limit > 0 && limit <= 20 ? limit : 5;

        return reportService.getTopSellingProducts(start, end, topLimit);
    }

    @GetMapping("/dashboard")
    public DashboardDto getDashboard() {
        return reportService.getDashboard();
    }
}
